compiler/src/main.rs

```diff
@@ -0,0 +1,208 @@
+use std::env;
+use std::ffi::{CStr, CString, OsStr};
+use std::fs::{self, File};
+use std::io::{self, Write};
+use std::mem::MaybeUninit;
+use std::path::{Path, PathBuf};
+use std::process::ExitCode;
+use std::str::{self, FromStr};
+
+use bpaf::Bpaf;
+
+mod error;
+use error::CompilerError;
+
+#[derive(Debug, Clone, Bpaf)]
+#[bpaf(options, version)]
+pub struct Options {
+    #[bpaf(external(output_type), optional)]
+    output_type: Option<OutputType>,
+
+    #[bpaf(long("linked"))]
+    /// Produce linked LLVM IR (in combination with --ll)
+    linked: bool,
+
+    #[bpaf(short('o'), argument("file"))]
+    /// Output path
+    output_path: Option<PathBuf>,
+
+    #[bpaf(positional("filename"))]
+    /// PTX file
+    ptx_path: String,
+}
+
+fn main() -> ExitCode {
+    main_core().map_or_else(
+        |e| {
+            eprintln!("Error: {}", e);
+            ExitCode::FAILURE
+        },
+        |_| ExitCode::SUCCESS,
+    )
+}
+
+fn main_core() -> Result<(), CompilerError> {
+    let opts = options().run();
+
+    let output_type = opts.output_type.unwrap_or_default();
+
+    if opts.linked && output_type != OutputType::LlvmIr {
+        println!("Warning: option --linked only makes sense when combined with --ll. Ignoring.");
+    }
+
+    let ptx_path = Path::new(&opts.ptx_path).to_path_buf();
+    check_path(&ptx_path)?;
+
+    let output_path = match opts.output_path {
+        Some(value) => value,
+        None => get_output_path(&ptx_path, &output_type)?,
+    };
+    check_path(&output_path)?;
```

Comment:
No reason to check output path. We should just create it (and if it fails during creation then expose that failure)